### PR TITLE
[AMORO-3087]: Optimizer Support Flink 1.20 (#3087)

### LINKED ISSUE #3087: [Improvement]: Optimizer Support Flink 1.20
### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/amoro/issues?q=is%3Aissue) and found no similar issues.

### What would you like to be improved?

Optimizer Support Flink 1.20,Flink has officially introduced full support for the standard [YAML 1.2 syntax](https://nightlies.apache.org/flink/flink-docs-release-1.20/docs/deployment/config/#flink-configuration-file) Amoro Support new flink **conf.yaml** and **flink-conf.yaml**

### How should we improve?

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Subtasks

_No response_

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by czy006**
## Why are the changes needed?

Close #3087.

## Brief change log

- Support FlinkOptimizerContainer With Flink 1.20
- Support new flink config with **config.yaml**

## How was this patch tested?

- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [x] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

- Does this pull request introduce a new feature? (no)
- If yes, how is the feature documented? (not documented)

### CODE REVIEW

**Comment by czy006** [comment]
image

File: README.md [hunk 1]
```diff
@@ -117,9 +117,9 @@ Amoro is built using Maven with JDK 8 and JDK 17(only for `amoro-mixed-format/am
 * Build and skip dashboard: `mvn clean package -Pskip-dashboard-build`
 * Build and disable disk storage, RocksDB will NOT be introduced to avoid memory overflow: `mvn clean package -DskipTests -Pno-extented-disk-storage`
 * Build with hadoop 2.x(the default is 3.x) dependencies: `mvn clean package -DskipTests -Phadoop2`
-* Specify Flink version for Flink optimizer(the default is 1.18.1): `mvn clean package -DskipTests -Dflink-optimizer.flink-version=1.15.4`
+* Specify Flink version for Flink optimizer(the default is 1.20.0): `mvn clean package -DskipTests -Dflink-optimizer.flink-version=1.20.0`
   * If the version of Flink is below 1.15.0, you also need to add the `-Pflink-optimizer-pre-1.15` parameter: `mvn clean package -DskipTests -Pflink-optimizer-pre-1.15 -Dflink-optimizer.flink-version=1.14.6`
-* Specify Spark version for Spark optimizer(the default is 3.3.3): `mvn clean package -DskipTests -Dspark-optimizer.spark-version=3.2.2`
+* Specify Spark version for Spark optimizer(the default is 3.3.3): `mvn clean package -DskipTests -Dspark-optimizer.spark-version=3.3.3`
```
Comment (huyuanfeng2018) [inline README.md:122]: We should not change spark related documents in this PR

File: amoro-ams/amoro-ams-server/src/main/java/org/apache/amoro/server/manager/FlinkOptimizerContainer.java [hunk 2]
```diff
@@ -80,25 +84,15 @@ public class FlinkOptimizerContainer extends AbstractResourceContainer {
   public static final String FLINK_HOME_PROPERTY = "flink-home";
   public static final String FLINK_CONFIG_PATH = "/conf";
   public static final String FLINK_CONFIG_YAML = "/flink-conf.yaml";
+  // flink version >= 1.20 use it first
+  public static final String NEW_FLINK_CONFIG_YAML = "/config.yaml";
```
Comment (huyuanfeng2018) [inline amoro-ams/amoro-ams-server/src/main/java/org/apache/amoro/server/manager/FlinkOptimizerContainer.java:88]:
```
FLINK_CONFIG_YAML="/config.yaml"
OLD_FLINK_CONFIG_YAML="/flink-conf.yaml"
```
WDYT？

File: amoro-ams/amoro-ams-server/src/main/java/org/apache/amoro/server/manager/FlinkOptimizerContainer.java [hunk 3]
```diff
@@ -266,15 +254,62 @@ protected String buildOptimizerStartupArgsString(Resource resource) {
...
   }
 
+  /**
+   * Copy from flink 1.20 GlobalConfiguration.flatten Utils
+   *
+   * @param config
+   * @param keyPrefix
+   * @return
+   */
+  private static Map<String, Object> flatten(Map<String, Object> config, String keyPrefix) {
```
Comment (huyuanfeng2018) [inline amoro-ams/amoro-ams-server/src/main/java/org/apache/amoro/server/manager/FlinkOptimizerContainer.java:293]: How about using  `GlobalConfiguration.loadConfiguration()` directly?

can start another  to do this

File: amoro-ams/amoro-ams-server/src/main/java/org/apache/amoro/server/manager/FlinkOptimizerContainer.java [hunk 4]
```diff
@@ -532,7 +563,7 @@ protected JobID runJar(String jarId, Configuration configuration, Resource resou
     JobID jobID = JobID.generate();
     JarRunRequestBody runRequestBody =
         new JarRunRequestBody(
-            FLINK_JOB_MAIN_CLASS, args, null, null, jobID, true, null, RestoreMode.DEFAULT, null);
+            FLINK_JOB_MAIN_CLASS, args, null, null, jobID, true, null, RestoreMode.CLAIM, null);
```
Comment (huyuanfeng2018) [inline amoro-ams/amoro-ams-server/src/main/java/org/apache/amoro/server/manager/FlinkOptimizerContainer.java:566]: Do we need to use RestoreMode.CLAIM? Essentially, Optimizer does not checkpoint, the default mode is NO_CLAIM

**Review by huyuanfeng2018 — COMMENTED** [review]
I leave some comments, PTAL

File: amoro-ams/amoro-ams-server/src/main/java/org/apache/amoro/server/manager/FlinkOptimizerContainer.java (hunk 3 above)
Comment (czy006) [reply amoro-ams/amoro-ams-server/src/main/java/org/apache/amoro/server/manager/FlinkOptimizerContainer.java:293]: This method does not exist in earlier versions of Flink, and **NoSuchMethodError** is raised when a user needs to compile Flink < 1.19

Comment (huyuanfeng2018) [reply amoro-ams/amoro-ams-server/src/main/java/org/apache/amoro/server/manager/FlinkOptimizerContainer.java:293]: I remember that this method already existed in a very early version of flink. After 1.19, config.yaml was used as the first priority configuration file.

File: amoro-ams/amoro-ams-server/src/main/java/org/apache/amoro/server/manager/FlinkOptimizerContainer.java (hunk 2 above)
Comment (huyuanfeng2018) [reply amoro-ams/amoro-ams-server/src/main/java/org/apache/amoro/server/manager/FlinkOptimizerContainer.java:88]: > > OLD_FLINK_CONFIG_YAML="/flink-conf.yaml"
> 
> I suggest using `LEGACY_FLINK_CONF_YAML`.

+1.

Comment (czy006) [reply amoro-ams/amoro-ams-server/src/main/java/org/apache/amoro/server/manager/FlinkOptimizerContainer.java]: LEGACY_FLINK_CONF_YAML +1

File: amoro-ams/amoro-ams-server/src/main/java/org/apache/amoro/server/manager/FlinkOptimizerContainer.java (hunk 3 above)
Comment (czy006) [reply amoro-ams/amoro-ams-server/src/main/java/org/apache/amoro/server/manager/FlinkOptimizerContainer.java:293]: It is worth noting that their implementation is completely different. There is new YAML file processing logic in the new version, so we may have inconsistent behavior on different versions

- [FLINK-1.15](https://github.com/apache/flink/blob/release-1.15.1/flink-core/src/main/java/org/apache/flink/configuration/GlobalConfiguration.java#L105)
- [FLINK-1.20](https://github.com/apache/flink/blob/release-1.20.0/flink-core/src/main/java/org/apache/flink/configuration/GlobalConfiguration.java#L99)

cc @huyuanfeng2018 @reswqa

Comment (czy006) [reply amoro-ams/amoro-ams-server/src/main/java/org/apache/amoro/server/manager/FlinkOptimizerContainer.java:293]: The current implementation is to copy this method from the high version of Flink for separate maintenance and compatibility with the low version of Flink, if this method is feasible I will continue to keep this part of the code WDYT？cc @reswqa @huyuanfeng2018

**Review by huyuanfeng2018 — APPROVED** [review]
LGTM.

**Comment by czy006** [comment]
I'll test with Flink1.18 later, no problem I'll merge

**Comment by zhoujinsong** [comment]
Thanks for the contribution! @czy006 

Do you plan to change the flink version for the amoro-flink-optimizer docker image?
You can find the related configuration here: https://github.com/apache/amoro/blob/master/.github/workflows/docker-images.yml#L100

**Comment by czy006** [comment]
> Thanks for the contribution! @czy006
> 
> Do you plan to change the flink version for the amoro-flink-optimizer docker image? You can find the related configuration here: https://github.com/apache/amoro/blob/master/.github/workflows/docker-images.yml#L100

I will test and change in this pr

**Comment by czy006** [comment]
> I'll test with Flink1.18 later, no problem I'll merge

image
image

**Comment by zhoujinsong** [comment]
Thanks for the contribution! @czy006 
Thanks for the review! @huyuanfeng2018 @reswqa

**Pull request merged**